### PR TITLE
FEAT: Create players shots

### PR DESCRIPTION
**Pull request by mathieurobert1**
I added the possibility to shoot with player.
image
I also fixed segfaults due to the vector of registries in SceneManager. Now the vector is removed by a single registry.

I also fixed the deletion of entities in the registry.

### CODE REVIEW

File: R-TypeGame/Components/Shoot/ShootComponent.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * @brief Component Shoot for R-Type Game.
+ *
+ */
+class ShootComponent : public AComponent {
+
+    public:
+
+        int damage;         // Damage of one bullet.
```
Comment (Thomaltarix) [inline R-TypeGame/Components/Shoot/ShootComponent.hpp:20]: Can damage be negative ?

File: R-TypeGame/Systems/MoveSystem/MoveEntities/MoveEntitiesSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        for (ECS::entity_t entity = 0; positions.size() >= entity + 1 && velocities.size() >= entity + 1; entity++) {
+            std::shared_ptr<Position2DComponent> position = std::dynamic_pointer_cast<Position2DComponent>(positions[entity]);
+            std::shared_ptr<VelocityComponent> velocity = std::dynamic_pointer_cast<VelocityComponent>(velocities[entity]);
+            if (!position || !velocity)
+                continue;
+            position->x += velocity->vx;
+            position->y += velocity->vy;
+        }
+    } catch (std::exception e) {
+    }
```
Comment (Thomaltarix) [inline R-TypeGame/Systems/MoveSystem/MoveEntities/MoveEntitiesSystem.cpp:32]: Why catching ?

File: R-TypeGame/Systems/Shoot/ShootDestroySystem.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+            if (position && shoot->friendlyFire) {
+                if (position->x >= 0 && position->x <= libGraphic->getWindowSize().first &&
+                    position->y >= 0 && position->y <= libGraphic->getWindowSize().second)
+                    continue;
+                reg.kill_entity(entity);
+            }
+        }
+    } catch (std::exception e){
+    }
```
Comment (Thomaltarix) [inline R-TypeGame/Systems/Shoot/ShootDestroySystem.cpp:44]: Catching for what ?

File: R-TypeGame/Systems/Shoot/ShootSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            std::shared_ptr<Position2DComponent> position = std::dynamic_pointer_cast<Position2DComponent>(positions[shoot]);
+            std::shared_ptr<Position2DComponent> positionPlayer = std::dynamic_pointer_cast<Position2DComponent>(positions[idxPacketEntities]);
+
+            if (position && positionPlayer) {
+                position->x = positionPlayer->x + POS_PLAYER_X;
+                position->y = positionPlayer->y + POS_PLAYER_Y;
+            }
+        }
+    } catch (std::exception e){
+    }
```
Comment (Thomaltarix) [inline R-TypeGame/Systems/Shoot/ShootSystem.cpp:36]: Catching for what ?

**Review by Thomaltarix — CHANGES_REQUESTED** [review]
Some questions about catches or variable types.
Also, for all your `*_registries`, you can replace it by `_registries.get()` if I'm not crazy.

Edit: I am crazy

File: R-TypeGame/Components/Shoot/ShootComponent.hpp (hunk 1 above)
Comment (mathieurobert1) [reply R-TypeGame/Components/Shoot/ShootComponent.hpp:20]: Perhaps if you enable healing with friendly shots ?

File: R-TypeGame/Systems/MoveSystem/MoveEntities/MoveEntitiesSystem.cpp (hunk 2 above)
Comment (mathieurobert1) [reply R-TypeGame/Systems/MoveSystem/MoveEntities/MoveEntitiesSystem.cpp:32]: I don't want to spam useless messages

File: R-TypeGame/Systems/Shoot/ShootDestroySystem.cpp (hunk 3 above)
Comment (mathieurobert1) [reply R-TypeGame/Systems/Shoot/ShootDestroySystem.cpp:44]: same

File: R-TypeGame/Systems/Shoot/ShootSystem.cpp (hunk 4 above)
Comment (mathieurobert1) [reply R-TypeGame/Systems/Shoot/ShootSystem.cpp:36]: same

File: R-TypeGame/Components/Shoot/ShootComponent.hpp (hunk 1 above)
Comment (Thomaltarix) [reply R-TypeGame/Components/Shoot/ShootComponent.hpp:20]: What a weird feature but okay 😂

File: R-TypeGame/Systems/MoveSystem/MoveEntities/MoveEntitiesSystem.cpp (hunk 2 above)
Comment (Thomaltarix) [reply R-TypeGame/Systems/MoveSystem/MoveEntities/MoveEntitiesSystem.cpp:32]: Fine

File: R-TypeGame/Components/Shoot/ShootComponent.hpp (hunk 1 above)
Comment (mathieurobert1) [reply R-TypeGame/Components/Shoot/ShootComponent.hpp:20]: No it's not, imagine a boost with which you can heal your friends during a short time.

Comment (Thomaltarix) [reply R-TypeGame/Components/Shoot/ShootComponent.hpp:20]: Yeah, why not, but I think that this feature should be implemented in the system that checks iff the shot hits an enemy or an ally, then apply `damage` to enemy or heal allies with `damage`

Comment (Thomaltarix) [reply R-TypeGame/Components/Shoot/ShootComponent.hpp:20]: But I'm fine with keeping damage as an integer 👍

**Review by Thomaltarix — APPROVED** [review]

**Review by Peralban — APPROVED** [review]

**Pull request merged**